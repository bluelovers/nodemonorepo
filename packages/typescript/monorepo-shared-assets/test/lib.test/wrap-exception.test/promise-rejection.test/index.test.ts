import { wrapException } from '../../../../index'
import wrap = wrapException.wrapPromiseRejection

namespace resolved {
  export function synchronous (value: number) {
    return new Result({ value })
  }

  export function asynchronous (value: number) {
    return Promise.resolve(new Result({ value }))
  }
}

namespace rejected {
  export function synchronous () {
    throw new Error('This is a rejection reason')
  }

  export function asynchronous () {
    return Promise.reject(new Error('This is a rejection reason'))
  }
}

namespace handleRejection {
  export function synchronous (reason: any, param: number) {
    return new Result({ reason, param })
  }

  export function asynchronous (reason: any, param: number) {
    return Promise.resolve(new Result({ reason, param }))
  }
}

class Result {
  constructor (dict: object) {
    Object.assign(this, dict)
  }
}

describe('with synchronous main', () => {
  describe('without rejection handler', () => {
    it('when promise resolves', async () => {
      const fn = wrap(resolved.synchronous)
      expect(await fn(1234)).toMatchSnapshot()
    })

    it('when promise rejects', async () => {
      const fn = wrap(rejected.synchronous)
      await expect(fn(1234)).rejects.toMatchSnapshot()
    })
  })

  describe('with synchronous rejection handler', () => {
    it('when promise resolves', async () => {
      const fn = wrap(resolved.synchronous, handleRejection.synchronous)
      expect(await fn(1234)).toMatchSnapshot()
    })

    it('when promise rejects', async () => {
      const fn = wrap(rejected.synchronous, handleRejection.synchronous)
      expect(await fn(1234)).toMatchSnapshot()
    })
  })

  describe('with asynchronous rejection handler', () => {
    it('when promise resolves', async () => {
      const fn = wrap(resolved.synchronous, handleRejection.asynchronous)
      expect(await fn(1234)).toMatchSnapshot()
    })

    it('when promise rejects', async () => {
      const fn = wrap(rejected.synchronous, handleRejection.asynchronous)
      expect(await fn(1234)).toMatchSnapshot()
    })
  })
})

describe('with asynchronous main', () => {
  describe('without rejection handler', () => {
    it('when promise resolves', async () => {
      const fn = wrap(resolved.asynchronous)
      expect(await fn(1234)).toMatchSnapshot()
    })

    it('when promise rejects', async () => {
      const fn = wrap(rejected.asynchronous)
      await expect(fn(1234)).rejects.toMatchSnapshot()
    })
  })

  describe('with synchronous rejection handler', () => {
    it('when promise resolves', async () => {
      const fn = wrap(resolved.asynchronous, handleRejection.synchronous)
      expect(await fn(1234)).toMatchSnapshot()
    })

    it('when promise rejects', async () => {
      const fn = wrap(rejected.asynchronous, handleRejection.synchronous)
      expect(await fn(1234)).toMatchSnapshot()
    })
  })

  describe('with asynchronous rejection handler', () => {
    it('when promise resolves', async () => {
      const fn = wrap(resolved.asynchronous, handleRejection.asynchronous)
      expect(await fn(1234)).toMatchSnapshot()
    })

    it('when promise rejects', async () => {
      const fn = wrap(rejected.asynchronous, handleRejection.asynchronous)
      expect(await fn(1234)).toMatchSnapshot()
    })
  })
})
