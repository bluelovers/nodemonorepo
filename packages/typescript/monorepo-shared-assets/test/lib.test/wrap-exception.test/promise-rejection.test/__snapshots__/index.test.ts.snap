// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with asynchronous main with asynchronous rejection handler when promise rejects 1`] = `
Result {
  "param": 1234,
  "reason": [Error: This is a rejection reason],
}
`;

exports[`with asynchronous main with asynchronous rejection handler when promise resolves 1`] = `
Result {
  "value": 1234,
}
`;

exports[`with asynchronous main with synchronous rejection handler when promise rejects 1`] = `
Result {
  "param": 1234,
  "reason": [Error: This is a rejection reason],
}
`;

exports[`with asynchronous main with synchronous rejection handler when promise resolves 1`] = `
Result {
  "value": 1234,
}
`;

exports[`with asynchronous main without rejection handler when promise rejects 1`] = `[Error: This is a rejection reason]`;

exports[`with asynchronous main without rejection handler when promise resolves 1`] = `
Result {
  "value": 1234,
}
`;

exports[`with synchronous main with asynchronous rejection handler when promise rejects 1`] = `
Result {
  "param": 1234,
  "reason": [Error: This is a rejection reason],
}
`;

exports[`with synchronous main with asynchronous rejection handler when promise resolves 1`] = `
Result {
  "value": 1234,
}
`;

exports[`with synchronous main with synchronous rejection handler when promise rejects 1`] = `
Result {
  "param": 1234,
  "reason": [Error: This is a rejection reason],
}
`;

exports[`with synchronous main with synchronous rejection handler when promise resolves 1`] = `
Result {
  "value": 1234,
}
`;

exports[`with synchronous main without rejection handler when promise rejects 1`] = `[Error: This is a rejection reason]`;

exports[`with synchronous main without rejection handler when promise resolves 1`] = `
Result {
  "value": 1234,
}
`;
